message.go

```diff
@@ -51,15 +51,15 @@ func (m *Message) decodeToNull() error {
 	return err
 }
 
-func (m *Message) makeDecodeNext(debugHook func(interface{})) DecodeNext {
+func (m *message) makeDecodeNext(debugHook func(interface{})) DecodeNext {
 	// Reserve the next object
-	m.t.ReadLock()
+	m.t.Lock()
 	return func(i interface{}) error {
 		ret := m.Decode(i)
 		if debugHook != nil {
 			debugHook(i)
 		}
-		m.t.ReadUnlock()
```

Comment:
I'm not familiar with this code, so can you explain a bit why you changed this to a write lock?

Also, btw, are you planning on changing this in a future PR? The unbalanced lock is setting off all kinds of alarm bells in my head...